apps/src/gamelab/GameLab.js

```diff
@@ -282,6 +285,99 @@ GameLab.prototype.reset = function (ignore) {
   this.executionError = null;
 };
 
+GameLab.prototype.onPuzzleComplete = function (submit) {
+  if (this.executionError) {
+    this.result = this.studioApp_.ResultType.ERROR;
+  } else {
+    // In most cases, submit all results as success
+    this.result = this.studioApp_.ResultType.SUCCESS;
+  }
+
+  // If we know they succeeded, mark levelComplete true
+  var levelComplete = (this.result === this.studioApp_.ResultType.SUCCESS);
+
+  if (this.executionError) {
+    this.testResults = this.studioApp_.getTestResults(levelComplete, {
+        executionError: this.executionError
+    });
+  } else if (!submit) {
+    this.testResults = this.studioApp_.TestResults.FREE_PLAY;
+  }
+
+  // Stop everything on screen
+  this.reset();
+
+  if (this.testResults >= this.studioApp_.TestResults.FREE_PLAY) {
+    this.studioApp_.playAudio('win');
+  } else {
+    this.studioApp_.playAudio('failure');
+  }
+
+  var program;
+
+  if (this.level.editCode) {
+    // If we want to "normalize" the JavaScript to avoid proliferation of nearly
+    // identical versions of the code on the service, we could do either of these:
+
+    // do an acorn.parse and then use escodegen to generate back a "clean" version
+    // or minify (uglifyjs) and that or js-beautify to restore a "clean" version
+
+    program = this.studioApp_.getCode();
+  } else {
+    var xml = Blockly.Xml.blockSpaceToDom(Blockly.mainBlockSpace);
+    program = Blockly.Xml.domToText(xml);
+  }
+
+  this.waitingForReport = true;
+
+  var sendReport = function() {
+    this.studioApp_.report({
+      app: 'gamelab',
+      level: this.level.id,
+      result: levelComplete,
+      testResult: this.testResults,
+      submitted: submit,
+      program: encodeURIComponent(program),
+      image: this.encodedFeedbackImage,
+      onComplete: (submit ? this.onSubmitComplete.bind(this) : this.onReportComplete.bind(this))
+    });
+
+    if (this.studioApp_.isUsingBlockly()) {
+      // reenable toolbox
+      Blockly.mainBlockSpaceEditor.setEnableToolbox(true);
+    }
+  }.bind(this);
+
+  var divGameLab = document.getElementById('divGameLab');
+  if (!divGameLab || typeof divGameLab.toDataURL === 'undefined') { // don't try it if function is not defined
+    sendReport();
+  } else {
+    divGameLab.toDataURL("image/png", {
+      callback: function(pngDataUrl) {
```

Comment:
This is cool!  So we actually upload a snapshot of the visualization when we finish the level?